cmake/modules/RootNewMacros.cmake

```diff
@@ -513,7 +513,51 @@ function (ROOT_CXXMODULES_APPEND_TO_MODULEMAP library library_headers)
   endforeach()
   #set(modulemap_entry "${modulemap_entry}  link \"lib/${library}\"\n")
   set(modulemap_entry "${modulemap_entry}  export *\n}\n\n")
-  set_property(GLOBAL APPEND PROPERTY ROOT_CXXMODULES_EXTRA_MODULEMAP_CONTENT ${modulemap_entry})
+  # Non ROOT projects need a modulemap generated for them in the current
+  # directory. The same happens with test dictionaries in ROOT which are not
+  # exposed via the main modulemap. This is exposed by setting the
+  # ROOT_CXXMODULES_WRITE_TO_CURRENT_DIR.
+  if (NOT "${CMAKE_PROJECT_NAME}" STREQUAL ROOT OR ROOT_CXXMODULES_WRITE_TO_CURRENT_DIR)
+    set(modulemap_output_file "${CMAKE_CURRENT_BINARY_DIR}/module.modulemap")
+
+    # It's possible that multiple modulemaps are needed in the current
+    # directory and we need to merge them. As we don't want to have multiple
+    # modules in the same moduluemap when rerunning CMake, we do a quick
+    # check if the current module is already in the modulemap (in which case
+    # we know we rerun CMake at the moment and start writing a new modulemap
+    # instead of appending new modules).
+
+    # The string we use to identify if the current module is already in the
+    # modulemap.
+    set(modulemap_needle "module \"${library}\"")
+    # Check if the needle is in the modulemap. If the file doesn't exis
+    # we just pretend we didn't found the string in the modulemap.
+    if (EXISTS "${modulemap_output_file}")
+      file(READ "${modulemap_output_file}" existing_contents)
+      string(FIND "${existing_contents}" "${modulemap_needle}" match_result)
+    else()
+      set(match_result -1)
```

Comment:
Can't you check if `match_result` is set? That would drop this branch and simplify the `if`-stmt.

Reply:
Hmm, I don't think that's a good idea. The problem is that if I set match_result in any scope of the stack that leads to this function call we will silently do the wrong thing then. E.g. in this scenario we will do the wrong thing:

```CMake
# Some random CMake file
string(FIND "${parameters}" "${my_parameter}" match_result)
if(match_result)
  # User has passed my_parameter, do something with it
endif()

# This call is now broken and sees match_result.
ROOT_GENERATE_DICTIONARY(...)
```
If you want to test this:
```CMake
project(Foo)
cmake_minimum_required(VERSION 3.8)

set(cool_var x)

function(foo)
  if(cool_var)
  message(STATUS "cool_var is set :O")
  endif()
endfunction()

# This will print x
message(STATUS "F: ${cool_var}")
```

Reply:
We could unconditionally reset match_result to -1 before the if, that would make it 1 line shorter and maybe a bit easier :)

Reply:
Ok, I see. Maybe that issue could be solved by checking if somebody defined already this. Either way I agree there is not a good way to simplify the code above.

Reply:
Ok, it's now a bit simpler.